pyteal/ast/itxn.py

```diff
@@ -200,6 +205,108 @@ def SetFields(cls, fields: Dict[TxnField, Union[Expr, List[Expr]]]) -> Expr:
         fieldsToSet = [cls.SetField(field, value) for field, value in fields.items()]
         return Seq(fieldsToSet)
 
+    @classmethod
+    def MethodCall(
+        cls,
+        *,
+        app_id: Expr,
+        method_signature: str,
+        args: list[abi.BaseType | Expr | dict[TxnField, Expr | list[Expr]]],
+        fields: dict[TxnField, Expr | list[Expr]],
+    ) -> Expr:
+
+        # Default, always need these
+        fields_to_set = [
+            cls.SetField(TxnField.type_enum, TxnType.ApplicationCall),
+            cls.SetField(TxnField.application_id, app_id),
+        ]
+
+        # We only care about the args
+        arg_type_specs: list[abi.TypeSpec]
+        arg_type_specs, _ = type_specs_from_signature(method_signature)
+
+        # Start app args with the method selector
+        app_args: list[Expr] = [MethodSignature(method_signature)]
+
+        # Transactions are not included in the App Call
+        txns_to_pass: list[Expr] = []
+
+        # Reference Types are treated specially
+        accts: list[Expr] = []
+        apps: list[Expr] = []
+        assets: list[Expr] = []
+
+        for idx, arg_ts in enumerate(arg_type_specs):
+            arg = args[idx]
+
+            if arg_ts in abi.TransactionTypeSpecs:
+                if not isinstance(arg, dict):
+                    raise TealTypeError(arg, dict[TxnField, Expr | list[Expr]])
+
+                txns_to_pass.append(InnerTxnBuilder.SetFields(arg))
+
+            elif arg_ts in abi.ReferenceTypeSpecs:
+                match arg_ts:
+                    # For both acct and application, add index to
+                    # app args _after_ appending since 0 is implicitly set
+                    case abi.AccountTypeSpec():
+                        if isinstance(arg, Expr):
+                            accts.append(arg)
+                        elif isinstance(arg, abi.Account):
+                            accts.append(arg.address())
+                        else:
+                            raise TealTypeError(arg, abi.Account | Expr)
+
+                        app_args.append(uint_encode(8, Int(len(accts))))
```

Comment:
`uint_encode` encodes the argument at runtime, which is less efficient. It would be better to encode the value in Python and wrap it in a `Bytes` expression.

To do the encoding I'd recommend doing something like `algosdk.abi.ABIType.from_string("uint8").encode(len(accts))`